File: cpp/data_structures/union_find.cc
#include <vector>

using namespace std;

class UnionFind {
 public:
  UnionFind(int n) : size(n, 1), parent(n, -1) {
    for (int i = 0; i < n; ++i) parent[i] = i;
  }

  bool IsJoined(int a, int b) {
    return Root(a) == Root(b);
  }

  void Join(int a, int b) {
    a = Root(a), b = Root(b);
    if (a == b) return;
    if (size[a] > size[b]) {
      size[a] += size[b];
      parent[b] = a;
    } else {
      size[b] += size[a];
      parent[a] = b;
    }
  }

 private:
  vector<int> size, parent;

  int Root(int a) {
    while (a != parent[a]) {
      parent[a] = parent[parent[a]];
      a = parent[a];
    }
    return a;
  }
};

File: python/data_structures/union_find.py

class UnionFind:
    def __init__(self, N):
        self.N = N
        self.parent = range(N)
        self.size = [0] * N

    def _root(self, a):
        while a != self.parent[a]:
            self.parent[a] = self.parent[self.parent[a]]
            a = self.parent[a]
        return a

    def is_joined(self, a, b):
        return self._root(a) == self._root(b)

    def join(self, a, b):
        a = self._root(a)
        b = self._root(b)
        if a == b:
            return
        if self.size[b] > self.size[a]:
            a, b = b, a  # Ensure a is the bigger tree
        self.size[a] += self.size[b]
        self.parent[b] = a


